src/transformers/clip.rs

```diff
@@ -240,37 +234,26 @@ impl<B: Backend> ClipAttention<B> {
     fn shape(&self, xs: Tensor<B, 3>, seq_len: usize, bsz: usize) -> Tensor<B, 4> {
         xs.reshape([bsz, seq_len, self.num_attention_heads, self.head_dim])
             .swap_dims(1, 2)
-        // .contiguous() // TODO: Figure out if this is needed or if we can abstract over memory
     }
 
-    fn forward(&self, xs: Tensor<B, 3>, causal_attention_mask: &Tensor<B, 2>) -> Tensor<B, 3> {
+    fn forward(&self, xs: Tensor<B, 3>, causal_attention_mask: Tensor<B, 4>) -> Tensor<B, 3> {
         let [bsz, seq_len, embed_dim] = xs.dims();
         let query_states = self.q_proj.forward(xs.clone()) * self.scale;
         let proj_shape = [bsz * self.num_attention_heads, seq_len, self.head_dim];
 
-        let query_states = self
-            .shape(query_states, seq_len, bsz)
-            .reshape(proj_shape)
-            .to_full_precision();
+        let query_states = self.shape(query_states, seq_len, bsz).reshape(proj_shape);
         let key_states = self
             .shape(self.k_proj.forward(xs.clone()), seq_len, bsz)
-            .reshape(proj_shape)
-            .to_full_precision();
+            .reshape(proj_shape);
         let value_states = self
             .shape(self.v_proj.forward(xs), seq_len, bsz)
-            .reshape(proj_shape)
-            .to_full_precision();
+            .reshape(proj_shape);
         let src_len = key_states.dims()[1];
         let attn_weights = query_states.matmul(key_states.swap_dims(1, 2));
 
         let attn_weights = attn_weights
             .reshape([bsz, self.num_attention_heads, seq_len, src_len])
-            .add(
-                causal_attention_mask
-                    .to_owned()
-                    .unsqueeze::<4>()
```

Comment:
mask is now a Tensor of shape 4, so we don't have to expand to that shape anymore

Reply:
where did the additional dimensions come from?

Reply:
In `build_casual_attention_mask` before it was a Tensor<B, 2> and now it's a Tensor<B, 4>

Reply:
Oh, I see